File: src/ia_dev_env/domain/pattern_mapping.py
from __future__ import annotations

from pathlib import Path
from typing import Dict, List

from ia_dev_env.models import ProjectConfig

UNIVERSAL_PATTERNS: List[str] = ["architectural", "data"]

ARCHITECTURE_PATTERNS: Dict[str, List[str]] = {
    "microservice": [
        "microservice",
        "resilience",
        "integration",
    ],
    "hexagonal": ["integration"],
    "monolith": ["integration"],
    "library": [],
}

EVENT_DRIVEN_PATTERNS: List[str] = [
    "saga-pattern",
    "outbox-pattern",
    "event-sourcing",
    "dead-letter-queue",
]


def select_patterns(config: ProjectConfig) -> List[str]:
    """Map architecture style to pattern category directories.

    Returns sorted, deduplicated list of pattern category names.
    Returns empty list for unknown styles.
    """
    style = config.architecture.style
    if style not in ARCHITECTURE_PATTERNS:
        return []
    categories = list(UNIVERSAL_PATTERNS)
    categories.extend(ARCHITECTURE_PATTERNS[style])
    if config.architecture.event_driven:
        categories.extend(EVENT_DRIVEN_PATTERNS)
    return sorted(set(categories))


def select_pattern_files(
    resources_dir: Path,
    pattern_categories: List[str],
) -> List[Path]:
    """List .md files from pattern category directories.

    Skips missing directories without error.
    Returns sorted list of file paths.
    """
    patterns_root = resources_dir / "patterns"
    files: List[Path] = []
    for category in pattern_categories:
        category_dir = patterns_root / category
        if not category_dir.is_dir():
            continue
        md_files = sorted(category_dir.glob("*.md"))
        files.extend(md_files)
    return files

File: src/domain/pattern-mapping.ts
/**
 * Pattern selection by architecture style.
 *
 * Migrated from Python `domain/pattern_mapping.py`.
 */

import type { ProjectConfig } from "../models.js";

import { readdirSync, statSync } from "node:fs";
import { join } from "node:path";

/** Universal patterns included for all known styles. */
export const UNIVERSAL_PATTERNS: readonly string[] = ["architectural", "data"];

/** Architecture style to pattern categories. */
export const ARCHITECTURE_PATTERNS: Readonly<Record<string, readonly string[]>> = {
  "microservice": ["microservice", "resilience", "integration"],
  "hexagonal": ["integration"],
  "monolith": ["integration"],
  "library": [],
};

/** Event-driven architecture patterns. */
export const EVENT_DRIVEN_PATTERNS: readonly string[] = [
  "saga-pattern",
  "outbox-pattern",
  "event-sourcing",
  "dead-letter-queue",
];

/**
 * Map architecture style to pattern category directories.
 *
 * Returns sorted, deduplicated list of pattern category names.
 * Returns empty list for unknown styles.
 */
export function selectPatterns(config: ProjectConfig): string[] {
  const style = config.architecture.style;
  const stylePatterns = ARCHITECTURE_PATTERNS[style];
  if (stylePatterns === undefined) {
    return [];
  }
  const categories = [...UNIVERSAL_PATTERNS, ...stylePatterns];
  if (config.architecture.eventDriven) {
    categories.push(...EVENT_DRIVEN_PATTERNS);
  }
  return [...new Set(categories)].sort();
}

/**
 * List .md files from pattern category directories.
 *
 * Skips missing directories without error.
 * Returns sorted list of file paths.
 */
export function selectPatternFiles(
  resourcesDir: string,
  patternCategories: readonly string[],
): string[] {
  const patternsRoot = join(resourcesDir, "patterns");
  const files: string[] = [];
  for (const category of patternCategories) {
    const categoryDir = join(patternsRoot, category);
    try {
      if (!statSync(categoryDir).isDirectory()) {
        continue;
      }
    } catch {
      continue;
    }
    const entries = readdirSync(categoryDir)
      .filter((f) => f.endsWith(".md"))
      .sort();
    for (const entry of entries) {
      files.push(join(categoryDir, entry));
    }
  }
  return files;
}
